src/Storages/MergeTree/CommonANNIndexes.h

```diff
@@ -0,0 +1,224 @@
+#pragma once
+
+#include <Storages/MergeTree/KeyCondition.h>
+#include <Storages/MergeTree/MergeTreeIndices.h>
+#include "base/types.h"
+
+#include <optional>
+#include <vector>
+
+namespace DB
+{
+
+namespace ANNCondition
+{
+
+/**
+    Class ANNCondition, is responsible for recognizing special query types which
+    can be speeded up by ANN Indexes. It parses the SQL query and checks
+    if it matches ANNIndexes. The recognizing method - alwaysUnknownOrTrue
+    returns false if we can speed up the query, and true otherwise.
+    It has only one argument, name of the metric with which index was built.
+    There are two main patterns of queries being supported
+
+    1) Search query type
+    SELECT * FROM * WHERE DistanceFunc(column, target_vector) < floatLiteral LIMIT count
+
+    2) OrderBy query type
+    SELECT * FROM * WHERE * ORDERBY DistanceFunc(column, target_vector) LIMIT count
+
+    *Query without LIMIT count is not supported*
+
+    target_vector(should have float coordinates) examples:
+        tuple(0.1, 0.1, ...., 0.1) or (0.1, 0.1, ...., 0.1)
+        [the word tuple is not needed]
+
+    If the query matches one of these two types, than the class extracts useful information
+    from the query. If the query has both 1 and 2 types, than we can't speed and alwaysUnknownOrTrue
+    returns true.
+
+    From matching query it extracts
+    * targetVector
+    * metricName(DistanceFunction)
+    * dimension size if query uses LpDistance
+    * distance to compare(ONLY for search types, otherwise you get exception)
+    * spaceDimension(which is targetVector's components count)
+    * column
+    * objects count from LIMIT section(for both queries)
+    * settings str, if query has settings section with new 'ann_index_params' value,
+        than you can get the new value(empty by default) calling method getSettingsStr
+    * queryHasOrderByClause and queryHasWhereClause return true if query matches the type
+
+    Search query type is also recognized for PREWHERE section
+*/
+
+class ANNCondition
+{
+public:
+    ANNCondition(const SelectQueryInfo & query_info,
+                    ContextPtr context);
+
+    // flase if query can be speeded up, true otherwise
+    bool alwaysUnknownOrTrue(String metric_name) const;
```

Comment:
This is a common thing for all indexes (skip, Ann) and KeyCondition. If indexes can't be applied (so it is Unknown) or always says that this granule should be used, it will be useful. indexIsUseful is more readable, but less informative, as we should guess, when it is useful. If it should be changed here, it should be changed everywhere